### PR TITLE
[#164] Improve Windows support

### LINKED ISSUE #164: Setup CI to run tests on Windows
# Clarification and motivation

**NB: we should probably do #165 first**

We've been building and [releasing](https://github.com/serokell/xrefcheck/releases/) windows binaries, but we haven't actually tested that they indeed work.

The readme also lists "Support for non-Unix systems" under the "For further work" section, so maybe we should invest some time into this.

The first thing to do is to setup a CI to run the haskell and bats tests on windows. GitHub Actions is probably the best way to set this up, see an example we can reuse [here](https://github.com/serokell/tztime/blob/336f585c2c7125a8ba58ffbf3dbea4f36a7c40e7/.github/workflows/ci.yml#L72-L118). The bats docs say [it supports Windows](https://bats-core.readthedocs.io/en/stable/installation.html), so this should be doable.

When this is up and running, we _may_ (hopefully not) find some bugs, in which case we'll open more follow-up issues.

A few things that _I think_ may be bugs:
* Backwards slashes in local links: `[a](b\c)`. Github and gitlab both interpret this as a link to a file named `b%5Cc`. But on a windows machine, will xrefcheck interpret this as a link to the file `c` in dir `b`?
* This line expects `rLink` to be an "absolute reference" and begin with a `/`. Appending it to `root` with `<>` works on linux, but will this behave correctly on windows? https://github.com/serokell/xrefcheck/blob/955e0ea9cf389e782cf30faf8f613756ea22cd0c/src/Xrefcheck/Verify.hs#L319

# Acceptance criteria

* We're using GitHub Actions to run the Haskell and Bats tests on Windows
* We've created issues for any bugs that come up

### PR DESCRIPTION
**Pull request by Sorokin-Anton**
## Description

When I tried to test `xrefcheck` on windows locally, I've found following problems:
-------------------------------------------------------------------------------
Problem: it requires some dynamic libraries to run, and `stack install` place  them in directory that is not on PATH by default
Solution: add this information and possible workarounds to README.md, create [issue is stack](https://github.com/commercialhaskell/stack/issues/5901)

-------------------------------------------------------------------------------
Problem: most Windows shells are not supporting utf8 by default. Sometimes printing of reports raise an error. `hPutChar: invalid argument (Invalid argument)`

Solution: use function `withCP65001` from `code-page` package which
sets correct codepage on Windows and do nothing on other OSs 

-------------------------------------------------------------------------------
Problem: in markdown links, '/' should always be used as path separator ,
e.g. GitHub renderer is not threating link `[x](a\b.md)` as a link
to file in dir `a`
We use OS-dependent file separators everywhere, so
e.g. `canonizeLocalRef "./a.md"="./a.md`"` on Windows

Solution: replace '\' to '/' while constructing repo tree if needed
and then use only '/', preferring functions from `System.FilePath.Posix`.
We can do this, since 'a/b' and `a\b` are equivalent paths on Windows

-------------------------------------------------------------------------------
Also I've added github actions workflow that runs `stack test xrefcheck:xrefcheck-tests` and golden tests  (this required modification of some tests)

## Related issue(s)

Fixes #164 

## :white_check_mark: Checklist for your Pull Request

Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](https://github.com/serokell/xrefcheck/tree/master/README.md)
    - Haddock

- Public contracts
  - [x] Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - [x] I added an entry to the [changelog](https://github.com/serokell/xrefcheck/tree/master/CHANGES.md) if my changes are visible to the users
        and
  - [x] provided a migration guide for breaking changes if possible

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](https://github.com/serokell/style/blob/master/haskell.md).

### CODE REVIEW

**Review by Sereja313 — APPROVED** [review]

**Comment by Sorokin-Anton** [comment]
Ok, now I need to think how to add case-sensetivity on Windows (maybe it will be separate issue)

File: CHANGES.md [hunk 1]
```diff
@@ -58,6 +62,7 @@ Unreleased
     If you've got `Expected a LINK after "ignore link" annotation` message, see
     PR's description for examples and details.
 
+
```
Comment (Martoon-00) [inline CHANGES.md:65]: Probably inserted by accident?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    - uses: actions/cache@v3
+      name: Cache AppData/Local/Programs/stack
+      with:
+        path: ~/AppData/Local/Programs/stack
+        key: ${{ runner.os }}-${{ matrix.ghc }}-appdata-stack
+
+
+# When editing this action, make sure it can run without using cached folders.
+# Yes, it tries to install mingw-w64-x86_64-pcre twice
```
Comment (Martoon-00) [inline .github/workflows/ci.yml:51]: Pacman does not feel too well? :troll:

**Review by Martoon-00 — COMMENTED** [review]
Looks good to me, leaving just a few minor comments.

I'll try to wrap my head around possible pitfalls for more and maybe will give another review later.

File: src/Xrefcheck/Command.hs [hunk 3]
```diff
@@ -47,7 +48,7 @@ findFirstExistingFile = \case
     if exists then pure (Just file) else findFirstExistingFile files
 
 defaultAction :: Options -> IO ()
-defaultAction Options{..} = do
+defaultAction Options{..} = withCP65001 $ do
```
[the remark this answers is not in the document]
Comment (Sorokin-Anton) [reply src/Xrefcheck/Command.hs:51]: Good idea. Created [issue ](https://github.com/serokell/haskell-with-utf8/issues/16) in that repo, also moved `withCP65001` to Main to keep such things in one place

File: README.md [hunk 4]
```diff
@@ -168,7 +168,17 @@ Currently supported options include:
...
 
 Run `stack install` to build everything and install the executable.
-If you want to use cabal, you need to run (`stack2cabal`)[https://hackage.haskell.org/package/stack2cabal] first!
+If you want to use cabal, you need to run [`stack2cabal`](https://hackage.haskell.org/package/stack2cabal) first!
+
+### Run on Windows [↑](#xrefcheck)
+On Windows, executable requires some dynamic libraries (DLLs).
+They are shipped together with executable in [releases page](https://github.com/serokell/xrefcheck/releases).
+If you have builded executable from source using `stack install`,
+those DLLs are downloaded by stack to location is not on `%PATH%` by default.
```
[the remark this answers is not in the document]
Comment (Sorokin-Anton) [reply README.md:177]: More English lessons... Thanks

File: tests/golden/check-ignoreExternalRefsTo/expected_windows.gold [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+     bad reference (external) at src:13:10-44:
+       - text: "labs"
+       - link: http://127.0.0.1:20000/labs
+       - anchor: -
+
+     ⛂  ConnectionFailure Network.Socket.connect: <socket: N>: failed (Connection refused (WSAECONNREFUSED))
+
+
+
+Invalid references dumped, 4 in total.
```
[the remark this answers is not in the document]
Comment (Sorokin-Anton) [reply tests/golden/check-ignoreExternalRefsTo/expected_windows.gold:35]: My bad, thanks

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Sorokin-Anton) [reply .github/workflows/ci.yml:51]: I just recorded my shell history, and then realized it's the way

File: CHANGES.md (hunk 1 above)
Comment (Sorokin-Anton) [reply CHANGES.md:65]: That is rebase-related, fixed

File: tests/Test/Xrefcheck/ConfigSpec.hs [hunk 6]
```diff
@@ -39,7 +39,9 @@ test_config =
     -- stack exec xrefcheck -- dump-config -t GitHub -o tests/configs/github-config.yaml
       [ testCase "Config matches" $ do
         config <- BS.readFile "tests/configs/github-config.yaml"
-        when (config /= defConfigText GitHub) $
+        -- On Windows, git clone can replace \n with \r\n in some files
+        let removeCR = BS.filter (/= '\r')
+        when (removeCR config /= removeCR (defConfigText GitHub)) $
```
[the remark this answers is not in the document]
Comment (Sorokin-Anton) [reply tests/Test/Xrefcheck/ConfigSpec.hs:44]: This is not a problem for yaml parser

**Comment by Sorokin-Anton** [comment]
Oh, looks like ci action fails because diff is now space-sensetive, so does not like cr-s. Locally  on Windows everything passes (no cr-s) , for me it's still a question where and why cr-s appear. Luckily there is diff option specially for that

File: src/Xrefcheck/Command.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Martoon-00) [reply src/Xrefcheck/Command.hs]: Awesome :+1:

File: tests/Test/Xrefcheck/ConfigSpec.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Martoon-00) [reply tests/Test/Xrefcheck/ConfigSpec.hs:44]: How about the user that will try to open and edit that file? Won't they see everything displayed in one line (at least in some text editors)?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Martoon-00) [reply .github/workflows/ci.yml:51]: It's just a bit of torture to see a mention of some magic without an explanation of why it is needed / why it works :sweat_smile: 

On the other hand, if all the possible explanation is "doing it once fails with X error, but somehow repeat helps both locally and in CI, no one knows why", then maybe it's really not necessary to provide.

**Comment by Martoon-00** [comment]
> Oh, looks like ci action fails because diff is now space-sensetive

I would take a more thorough look at this, as it may point to a real UX issue of xrefcheck that will rise for some users.

I have two thoughts: 

1. AFAIR when I installed git in Windows, it asked me which line ending I want to use and whether do I want to replace non-matching line endings automatically.

    And looks like GitHub actions for Windows [have this conversion enabled by default](https://github.com/actions/checkout/issues/135). People in that ticket also mention issues with testing, so given that this problem is common for tests, disabling automatic line endings conversion can be a way to go.

    If this is an issue in our case, I don't have strong opinion on where to fix it, in `assert_diff` or in the ci config via overriding the git settings. 

2. In `xrefcheck`'s output, do we have cr-s or not?

    I suspect we need them, as otherwise not all terminals / editors (in case the user re-directs the output to a file) will manage to display that properly?
    
    On the other side, cr-s should already be included because TIL in Haskell's `base` package input/output streams are [automatically updated](https://downloads.haskell.org/~ghc/8.0.1/docs/html/libraries/base-4.9.0.0/System-IO.html#g:25) to handle the newline endings :thinking: 
    
However judging by your symptops (as far as I got them correctly), I completely don't understand which of these points currently take place and which may need an adjustment.

**Comment by Sorokin-Anton** [comment]
@Martoon-00 finally, I've got two results:
If I gather xrefcheck's output via `xrefcheck.exe > tmpfile`, there are `\r`s
If I run `xrefcheck.exe dump-config -t GitHub`, thee are Unix line endings (no `\r`s)
I will try to use a different function so the config would be dumped with `\r`s. 
About diffs - I'd prefer to use checkout with `\r`s since this would also allow us to simplify one Haskell test.

**Comment by Sorokin-Anton** [comment]
@Martoon-00 Bad news: if the repository was cloned with crlf's, output files have extra caret returns e.g. `=== Invalid references found ===\r\r\n\r\r\n`
So I need some time to handle this

**Comment by Sorokin-Anton** [comment]
The reason of this is
```
ghci> TL.unlines ["a","b"]        
"a\nb\n"
```
and
```
ghci> TL.lines "a\r\nb\r\nc"
["a\r","b\r","c"]
```
while nyan-interpolation leaves \r's

**Comment by Martoon-00** [comment]
Awesome. Very nice investigation!

I agree with your conclusions. Quite sad that this was produced by nyan-interpolation (suddenly, initially I would never suspect it). We will have to get it fixed there.

**Comment by Sorokin-Anton** [comment]
@martoon interesting fact: to_temp from bats always creates a file with LF endings 😐
I can't see why, looks like it is not a problem with our helpers. Same problem with assert_output
Looks like it's impossible to run assert_output --partial on multiline output on WIndows (and we use this to check error message in one case).
So I'm leaving `--strip-trailing-cr` in `assert_diff ` and rewriting this one check somehow.

**Comment by Martoon-00** [comment]
Meeh, I see.

Your solution to this sounds good. I have no other concerns.

---

Just a few words on this farewell I guess: that was a pleasure to review the work of yours. In my feeling, it required fewer review iterations than on average as you did everything mostly perfectly from the beginning. It was comfortable to me how initiative you were and has far you were ready to go when new issues arose. And concerns with English were pretty minor IMO.

**Review by Martoon-00 — APPROVED** [review]

**Pull request merged**